### ISSUE TITLE
Keypoint loss too relaxed for custom number of keypoints

### ISSUE THREAD

**Issue by davyneven** [ultralytics/ultralytics#2543]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Training

### Bug

I'm playing around with your pose/keypoint training script on custom keypoint datasets but stumbled upon a few issues.

I noticed that the keypoint metrics (mAP50/mAP50-95) are way too high when comparing the location of predicted keypoints with the groundtruth. I am having an mAP50-95 close to 99, while visually the keypoints on the training set are still way of their target. 

By going through the codebase, I might have found what is causing this (and how to solve it):

First of all, I believe the sigmas for custom keypoints are not correctly initialized (too forgiving). Currently they are set to 
`torch.ones(nkpt, device=self.device) / nkpt`  where I believe you forgot to add a `/ 10`.

For example, for nkpt = 2, this would lead to a sigma of 0.5 for each keypoint, which is way too forgiving. For example, given a bbox of 100 by 200 pixels, an offset of 20 pixels would still yield a OKS of 0.99 ...

This leads to two issues: first the obvious one is that the metrics are way too positive, but more importantly, the loss is too relaxed, so the predicted keypoints don't move any longer when they are already in (not so ) close proximity to the keypoint.

One way to fix the metric would indeed be to lower the sigma value, to for example 0.05 (instead of adding  `/ 10`, I would believe a fixed value for each keypoint would be better). However, as you are using the OKS metric as a loss function, this leads to very low gradients (as the derivative of the gaussian is almost zero when too far off, which is the case at the beginning of training. This might be addressed by better initialization, but too difficult to tune for any custom dataset. In my case, after 100 epochs, I still had an mAP50 of 0 ...

I believe there are two options to solve this issue:

1. Using a different loss function. I managed to get very good results using the standard L1-loss (and lowering the pose weight to 5 instead of 12):

```
class KeypointLoss(nn.Module):

    def __init__(self, sigmas) -> None:
        super().__init__()
        self.sigmas = sigmas

    def forward(self, pred_kpts, gt_kpts, kpt_mask, area):
        d = (pred_kpts[..., 0] - gt_kpts[..., 0]).abs() + (pred_kpts[..., 1] - gt_kpts[..., 1]).abs()
        kpt_loss_factor = (torch.sum(kpt_mask != 0) + torch.sum(kpt_mask == 0)) / (torch.sum(kpt_mask != 0) + 1e-9)
        return kpt_loss_factor * (d * kpt_mask).mean()
```

2. What also works is a combination of the two, which might be best, so you are still directly optimizing the OKS metric (like how the combine BCE and dice loss for semantic segmentation):

```
        d = (pred_kpts[..., 0] - gt_kpts[..., 0]) ** 2 + (pred_kpts[..., 1] - gt_kpts[..., 1]) ** 2
        d_abs = d = (pred_kpts[..., 0] - gt_kpts[..., 0]).abs() + (pred_kpts[..., 1] - gt_kpts[..., 1]).abs()
        kpt_loss_factor = (torch.sum(kpt_mask != 0) + torch.sum(kpt_mask == 0)) / (torch.sum(kpt_mask != 0) + 1e-9)
        # e = d / (2 * (area * self.sigmas) ** 2 + 1e-9)  # from formula
        e = d / (2 * self.sigmas) ** 2 / (area + 1e-9) / 2  # from cocoeval
        return kpt_loss_factor * (((1 - torch.exp(-e))*0.1 + d_abs*0.9)* kpt_mask).mean()
```

So, to cut my story short, I believe that to train a custom keypoint task, the default OKS loss fails, and you have to reside to the standard L1/L2 loss or a combination of the two. 

You can download my custom (white board markers) dataset here: https://drive.google.com/file/d/1csR0EkzGb3EQ1MEEvlXfsyVF2pbqUUw9/view?usp=sharing

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by Laughing-q** [comment]
@davyneven Thanks for the nice suggestions! Actually I tested a facial landmark dataset -- widerface to verify the generality of the loss function and after a few epochs the predictions looks good, since widerface does not provide keypoint info in their val set so I don't really know how the metrics look like. 
![987](https://github.com/ultralytics/ultralytics/assets/61612323/73e9d54c-d949-4149-bcfb-0991742c26e9)
And in my experiments the pose part convergence is obviously slower than the bbox part which usually needs more dataset and more epochs. 
Yes I agree you that the current loss function is not very suitable for some cases that only got 2, 3 keypoints. Could you make a PR then we can actually experiment on it? Thanks!

**Comment by CalinLucian** [comment]
I can also confirm this observation (I wanted to actually ask a different question on this but then again it's already opened).

**Comment by glenn-jocher** [comment]
@CalinLucian thank you for bringing this up! It is true that the bounding box mAP metric in YOLOv8 tends to converge faster compared to the keypoint metric in some cases. This is due to several factors, such as the complexity of the task, the amount and quality of the data, and the learning rate. Since the keypoint task is usually more complex than the bounding box task, it might require more data and longer training to converge. Additionally, the learning rate might need to be adjusted to stabilize the training process and help the keypoint metric to converge faster. 

In any case, it is important to keep monitoring both metrics during training to ensure that the model is performing well on both tasks. A good strategy is to use a combination of loss functions that can balance the importance of the two tasks, and to adjust the learning rate and other hyperparameters based on the feedback from the metrics. We appreciate your feedback and will consider ways to improve the training process for keypoint tasks in future updates.

**Comment by TimbusCalin** [comment]
@glenn-jocher Can I use a custom metric (such as Pose R/Pose P) as an optimizing metric when saving the best model for posing? Otherwise I think by default the one used is mAP50-95 for the BBox which is not helpful given different points in times in the convervence.

In addition, how can I use a custom loss function?

Thank you kindly for your time.

**Comment by LiangYong1216** [comment]
hi @davyneven ,what's  this means:(and lowering the pose weight to 5 instead of 12). 
So your idea is to change pose=12 to pose=5 during training? Is my understanding correct that the code should be model.train(data='header.yaml', epochs=1000, imgsz=640, batch=128, device="1, 2", dropout=0.5, pose=5)? all right?

**Comment by davyneven** [comment]
@LiangYong1216 indeed, as @glenn-jocher indicates, you can set pose=5 in the train function. However, this is only necessary if you also change the loss function, as I mentioned above; if you don't change the loss function, there is no need for lowering the pose weight.

**Comment by LiangYong1216** [comment]
hello @davyneven @glenn-jocher ,Thanks for your prompt reply, I set pose=5 on my dataset, but it doesn't seem to work, after 30 epochs all P and R are 1 during training, the loss function uses your second scheme, the model I use yolov8s-pose.pt. Each image in my dataset has 52 key points, and the markup is definitely correct, the model seems to be good, but there is still a certain gap compared to the label, how can I improve it?

**Comment by rm-ra** [comment]
> 你好[@LiangYong1216](https://github.com/LiangYong1216)感谢您的提问。 该代码`pose=12`用于定义训练期间姿势/关键点损失的权重因子。默认情况下，姿势权重因子设置为 12，这意味着姿势损失对总损失的贡献将大于边界框和分类损失。 在[@davyneven](https://github.com/davyneven)他们发现，将姿势权重因子降低到 5 可以使关键点度量更好地收敛。 要实现此更改，您可以`pose`按如下方式修改训练命令中的参数：`model.train(data='header.yaml', epochs=1000, imgsz=640, batch=128, device="1, 2", dropout=0.5, pose=5)`。这会在训练期间将姿势损失的权重因子设置为 5。 希望这能对您有所帮助！如果您有任何其他问题，请随时提问。
When I was training faces and five key points, the map50 of the first epoch reached 0.99. I tried using the above L1 loss and pose=5, but the map50 only reached 0.96. The dataset includes bigface, scenes, and widerface, with more than 70,000 images. Is this overfitting or a problem with the dataset?
![Image](https://github.com/user-attachments/assets/75e2443f-f64d-4eb2-b3cd-284e4882c780)

`task: pose
mode: train
model: yolov8s-pose.pt
data: yolo_config.yaml
epochs: 100
time: null
patience: 100
batch: 16
imgsz: 640
save: true
save_period: -1
cache: false
device: cuda:0
workers: 8
project: null
name: train3
exist_ok: false
pretrained: true
optimizer: auto
verbose: true
seed: 0
deterministic: true
single_cls: false
rect: false
cos_lr: false
close_mosaic: 10
resume: false
amp: true
fraction: 1.0
profile: false
freeze: null
multi_scale: false
overlap_mask: true
mask_ratio: 4
dropout: 0.0
val: true
split: val
save_json: false
conf: null
iou: 0.7
max_det: 300
half: false
dnn: false
plots: true
source: null
vid_stride: 1
stream_buffer: false
visualize: false
augment: false
agnostic_nms: false
classes: null
retina_masks: false
embed: null
show: false
save_frames: false
save_txt: false
save_conf: false
save_crop: false
show_labels: true
show_conf: true
show_boxes: true
line_width: null
format: torchscript
keras: false
optimize: false
int8: false
dynamic: false
simplify: true
opset: null
workspace: null
nms: false
lr0: 0.01
lrf: 0.01
momentum: 0.937
weight_decay: 0.0005
warmup_epochs: 3.0
warmup_momentum: 0.8
warmup_bias_lr: 0.1
box: 7.5
cls: 0.5
dfl: 1.5
pose: 5
kobj: 1.0
nbs: 64
hsv_h: 0.015
hsv_s: 0.7
hsv_v: 0.4
degrees: 20
translate: 0.1
scale: 0.5
shear: 0.0
perspective: 0.0
flipud: 0.0
fliplr: 0.2
bgr: 0.0
mosaic: 0.5
mixup: 0.0
cutmix: 0.0
copy_paste: 0.0
copy_paste_mode: flip
auto_augment: randaugment
erasing: 0.4
cfg: null
tracker: botsort.yaml
save_dir: runs/pose/train3`

**Comment by glenn-jocher** [comment]
Hi @rm-ra,

The drop in mAP50 from 0.99 to 0.96 when switching to L1 loss is actually expected and likely indicates the metrics are now more realistic. The artificially high mAP50 of 0.99 with the default loss was precisely the issue @davyneven identified - the default OKS-based loss with too-forgiving sigmas makes the metrics unrealistically high, especially for datasets with few keypoints.

Since your dataset contains many faces where keypoints cannot be marked (using -1 coordinates), this could be affecting your metrics calculation. The visibility values being only 0 and 2 (missing the intermediate visibility state of 1) might also impact how loss is calculated during training.

To better evaluate your model's performance:
1. Visually inspect predictions on a validation set to verify actual keypoint accuracy
2. Consider creating a clean test subset with accurate keypoints for proper evaluation
3. Use a custom metric that better reflects your use case requirements

For facial keypoints specifically, you might want to adjust the L1 loss implementation to handle the visibility flags properly or create a custom loss function that better suits your dataset characteristics.
